### PR TITLE
feat!: represent small values as single bytes

### LINKED ISSUE #1398: Representation of small values as single bytes
This is a preliminary task in preparation for the new encoding/decoding scheme.

We have other PRs to serve as a reference:
- https://github.com/FuelLabs/fuels-rs/pull/1163
- https://github.com/FuelLabs/sway/pull/4929

### PR DESCRIPTION
**Pull request by danielbate**
- Closes #1398

# Status

```
Test Suites: 250 passed, 250 total
Tests:       17 skipped, 1422 passed, 1439 total
```

# New Test Scenarios

New isolated tests are being done at:
 - `packages/fuel-gauge/src/small-bytes.test.ts`

# Failing tests

 - [x] @nedsalk `apps/docs-snippets/src/guide/contracts/configurable-constants.test.ts`
 - [x] @danielbate `apps/docs-snippets/src/guide/scripts/script-with-configurable.test.ts`
 - [x] `apps/docs-snippets/src/guide/types/options.test.ts`
 - [x] `apps/docs-snippets/src/guide/types/tuples.test.ts`
 - [x] @nedsalk `apps/docs-snippets/src/guide/types/vector.test.ts`

<br />

 - [x] @nedsalk `packages/abi-coder/test/interface.test.ts`

<br />

 - [x] @nedsalk `packages/fuel-gauge/src/advanced-logging.test.ts`
 - [x] @danielbate `packages/fuel-gauge/src/configurable-contract.test.ts`
 - [x] `packages/fuel-gauge/src/contract.test.ts`
 - [x] `packages/fuel-gauge/src/coverage-contract.test.ts`
 - [x] @danielbate `packages/fuel-gauge/src/predicate/predicate-configurables.test.ts`
 - [x] @Torres-ssf  `packages/fuel-gauge/src/predicate/predicate-with-contract.test.ts`
 - [x] @danielbate `packages/fuel-gauge/src/script-with-configurable.test.ts`
 - [x] `packages/fuel-gauge/src/script-with-vectors.test.ts`
 - [x] `packages/fuel-gauge/src/vector-types.test.ts`
 - [x] @nedsalk `packages/fuel-gauge/src/vectors.test.ts`

# Internal Docs

This is the summary of the new encoding/decoding scheme:

https://github.com/FuelLabs/fuels-ts/blob/de19b90b57e3d083a71d3b9464eaf7a65b35650b/packages/abi-coder/src/coders/abstract-coder.ts#L9-L47

### CODE REVIEW

File: packages/abi-coder/src/coders/struct.ts [hunk 1]
```diff
@@ -40,7 +41,12 @@ export class StructCoder<TCoders extends Record<string, Coder>> extends Coder<
           `Invalid ${this.type}. Field "${fieldName}" not present.`
         );
       }
-      const encoded = fieldCoder.encode(fieldValue);
+      let encoded = fieldCoder.encode(fieldValue);
+
+      if (encoded.length % 8 !== 0) {
+        encoded = concatBytes([encoded, new Uint8Array(8 - (encoded.length % 8))]);
+      }
+
```
Comment (arboleya) [inline packages/abi-coder/src/coders/struct.ts:49]: @nedsalk Good one! Could you please add detailed notes for these changes?

File: packages/abi-coder/src/coders/struct.ts [hunk 2]
```diff
@@ -53,6 +59,10 @@ export class StructCoder<TCoders extends Record<string, Coder>> extends Coder<
       const fieldCoder = this.coders[fieldName];
       let decoded;
       [decoded, newOffset] = fieldCoder.decode(data, newOffset);
+      if (newOffset % 8 !== 0) {
+        newOffset += 8 - (newOffset % 8);
+      }
+
```
Comment (arboleya) [inline packages/abi-coder/src/coders/struct.ts:65]: @nedsalk Same thing here; keeping precise notes of these inner changes will be very helpful.

File: apps/docs-snippets/src/guide/types/vector.test.ts [hunk 3]
```diff
@@ -36,13 +36,12 @@ describe(__filename, () => {
         age: 31,
         salary: 9000,
         idHash: getRandomB256(),
-        ratings: [1, 2, 3],
-        isActive: false,
+        ratings: [5, 6, 7],
+        isActive: true,
```
Comment (arboleya) [inline apps/docs-snippets/src/guide/types/vector.test.ts:40]: @nedsalk Considering the previous values should work just fine, why was this change necessary?

Comment (nedsalk) [reply apps/docs-snippets/src/guide/types/vector.test.ts:40]: It was for debugging purposes, because the array was the place where the underlying issue happened. I had some console.logs which logged the encoded `ratings` and `isActive` properties, but they were the same in both structs of the test so I changed them for easier discernment of which struct I'm looking at. I think it's better to leave it like this.

**Review by arboleya — CHANGES_REQUESTED** [review]
> [!WARNING]
> This PR should not be merged yet.
> Please, don't resolve this conversation.

**Comment by danielbate** [comment]
Closing as this work is included in #1407 so the source of truth review should be there
